### ISSUE TITLE
Implement smart pointer (de)serialization

### ISSUE THREAD

**Issue by switch-blade-stuff** [switch-blade-stuff/sekhmet#2]
Archives should support smart pointers (and other pointer-like object types). For this reason, every archive should have a pointer table used to ensure uniqueness of smart pointers.

This table would store the actual value of the smart pointer, `type_info` of the pointed-to type and a proxy serialization function used to later serialize the pointed-to type.

When a smart pointer is serialized, its existance is first checked within that table.
If it is present within the table, and the pointed-to types are different, the types are checked for inheritance.
When the pointer is not present within the table, or the pointed-to type is more derived than the existing entry, a type-erased copy of the pointer is stored within the table, and a proxy serialization function is created. Every entry is assigned a UUID which is then used to reference the pointed-to type. This UUID is what is stored when a smart pointer is serialized. Nil UUID represents a null smart pointer.
Once serialization of the main data is finished, every stored pointer entry is serialized in a table of UUID to the object type & data.

On de-serialization first the pointer table is read & de-serialized. Stored type info is used to allocate & initialize the serialized object.
Every object is then accessible through an internal table indexed by smart pointer UUIDs. When a smart pointer is de-serialized, it's UUID is used to get an object from that table, 

For the type to be constructible at runtime, it's type_info is used to get the default in-place constructor & destructor, size and alignment. To add pointer-like serialization support for custom types, the `template<typename PtrT> PtrT sek::pointer_like_factory(type_info)` function should be specialized for that type. This factory is used to initialize an instance of the smart pointer during de-serialization from a type info.

Pointer-like types that do not have a specialization of `pointer_like_factory` will be not support smart pointer serialization.

**Comment by switch-blade-stuff** [comment]
Scratch this. Builtin deferred pointer serialization isnt the best idea since it is a hidden cost & functionality.
Instead there is not ability to pass through arguments to (de)serialization functions, which can be used to implement pointer serialization.
